src/dataset.jl

```diff
@@ -0,0 +1,165 @@
+"""
+    Dataset{T}
+
+A `Dataset` describes an associative collection of component `KeyedArray`s with constraints
+on their shared dimensions.
+
+# Fields
+- `constraints::OrderedSet{Pattern}` - Constraint [`Pattern`](@ref)s on shared dimensions.
+- `data::LittleDict{Tuple{Vararg{Symbol}}, T}` - Flattened key paths as tuples of symbols
+  to each component array of type `T`.
+"""
+struct Dataset{T<:XArray}
```

Comment:
would it be useful to make this a Dict subtype as I see we are instantiating Dict iterators for this?

Reply:
I'm not sure. AFAICT, several types implement these methods without subtyping `AbstractDict`. I feel like we're deviating enough from the default behaviour that it might not make sense.

Pros:
- We could dispatch on `AbstractDict` and use this interchangeably, though we also often use `NamedTuple`.
- We might get some extra default functionality from that, though I doubt that default behaviour will correctly out of the box.

Cons:
- We risk allowing incorrect behaviour by falling back to `AbstractDict` methods
- We restrict what we can do with this type based on the fact that it's a dict